### PR TITLE
Circulars search backend for date range

### PR DESCRIPTION
**Pull request by tylerbarna**
Backend for Issue #878 

Date range filter has been implemented inside circulars.server.ts with the addition of an advancedQuery optional argument, which contains optional start and end date arguments that will default to negative and positive infinity, respectively. When the frontend is added in a separate PR, react-uswds has a date range picker element that can be used.

Added functions to circulars.lib.ts for converting a unix time to mm-dd-yy format and another function to do the opposite. The URLSearchParams for the start and end dates are in unix time (the circular items store the createdOn time in unix time format). In the search function, the iso8061 format dates are converted to unix times to simplify the filter. However, as Leo said in the 06/12/23 GCN meeting, it would be better if the date range values were stored in the iso8061 format in the url so they're human readable. This would require changes to index.tsx, but the formatDateEpoch implementation in the search function shouldn't require any changes since it should be able to convert to unix time if provided an iso8061 string.

Currently, the changes made in index.tsx to incorporate the date range parameters cause an unintended issue with the way URLSearchParams are retreived by the loader function after initiating a search; in the loader function, it seems that the only parameter retreived is the query parameter. The lack of the page number parameter causes the query results to not display correctly, though logging the results variable does show that they're being loaded, so I don' think it's due to the search function. Inspecting the url parameters in other functions, which print in the client-side console, seem to have all of the expected parameters. This is explained in slightly more detail in [this comment](https://github.com/nasa-gcn/gcn.nasa.gov/issues/878#issuecomment-1583244451) in the related issue.

### CODE REVIEW

**Comment by tylerbarna** [comment]
> Backend for Issue #878
> 
> Date range filter has been implemented inside circulars.server.ts with the addition of an advancedQuery optional argument, which contains optional start and end date arguments that will default to negative and positive infinity, respectively. When the frontend is added in a separate PR, react-uswds has a date range picker element that can be used.
> 

My thoughts on the frontend was to add an advanced query button (something like [this](https://github.com/tylerbarna/gcn.nasa.gov/blob/68391db7acf37ace4e5e4479379e90e54673da4a/app/routes/circulars/index.tsx#LL206C3-L206C3)) to the search bar that expands an accordion menu ([react-uswds documentation](https://trussworks.github.io/react-uswds/?path=/docs/components-accordion--borderless)) with different advanced query options, the current addition being a date range picker ([react-uswds documentation](https://trussworks.github.io/react-uswds/?path=/docs/components-date-range-picker--complete-date-range-picker))

File: app/routes/circulars/circulars.lib.ts [hunk 1]
```diff
@@ -76,6 +76,25 @@ export function formatDateISO(date: number) {
...
+  return dateStr.replace(/\//g, '-') // replace / with -
+}
+
+/** convert a date in format mm-dd-yyyy to ms since 01/01/1970 */
+export function formatDateEpoch(date: string | number) {
+  if (typeof date === 'string') {
+    return new Date(date).getTime()
+  }
+  return date
+}
```
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:96]: Use the format `YYYY-MM-DD`, not `MM-DD-YYYY`, to more closely match the international ISO 8601 format.

File: app/routes/circulars/circulars.lib.ts [hunk 2]
```diff
@@ -76,6 +76,25 @@ export function formatDateISO(date: number) {
...
+  })
+  return dateStr.replace(/\//g, '-') // replace / with -
+}
+
+/** convert a date in format mm-dd-yyyy to ms since 01/01/1970 */
+export function formatDateEpoch(date: string | number) {
+  if (typeof date === 'string') {
+    return new Date(date).getTime()
+  }
+  return date
```
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:95]: I think it's a bit too fancy for it to take either a UNIX timestamp or a string. Let's just have this take a string, so that this function is simply the inverse of `formatDateMMDDYYYY `.

```suggestion
export function formatDateEpoch(date: string) {
  return new Date(date).getTime()
```

File: app/routes/circulars/circulars.server.ts [hunk 3]
```diff
@@ -48,17 +49,29 @@ export async function search({
...
   limit,
+  advancedQuery,
 }: {
   query?: string
   page?: number
   limit?: number
+  advancedQuery?: {
+    startDate?: string | number
+    endDate?: string | number
+  }
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:60]: Move these out of the `advancedQuery` object:

```suggestion
  startDate?: string | number
  endDate?: string | number
```

File: app/routes/circulars/circulars.server.ts [hunk 4]
```diff
@@ -69,11 +82,27 @@ export async function search({
...
+                gte: startTime,
+                lte: endTime,
+              },
+            },
+          },
+        },
       },
-      fields: ['subject'],
-      _source: false,
+      _source: ['subject', 'createdOn'],
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:105]: This was probably a typo. I think that the `_source` argument is supposed to be a boolean.

File: app/routes/circulars/circulars.server.ts [hunk 5]
```diff
@@ -107,13 +138,12 @@ export async function search({
...
 /** Get a circular by ID. */
 export async function get(circularId: number): Promise<Circular> {
-  if (isNaN(circularId)) throw new Response(null, { status: 404 })
   const db = await tables()
   const result = await db.circulars.get({
     circularId,
   })
   if (!result)
-    throw new Response(null, {
+    throw new Response('The requested circular does not exist', {
```
Comment (lpsinger) [inline app/routes/circulars/circulars.server.ts:146]: Whoops, looks like this patch undid some unrelated changes. Possibly a merge artifact?

**Review by lpsinger — CHANGES_REQUESTED** [review]
I'm not sure of the cause of the bug that you are seeing, but I would suggest not storing complicated composite objects with `useState` because that is an easy source of bugs.

File: app/routes/circulars/circulars.lib.ts (hunk 2 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.lib.ts:95]: the rationale I had behind this was so the search function was more flexible when defining the startTime and endTime constants so it could intake either the unix format as a number or the human-readable/iso8601 format as a string (see [here](https://github.com/tylerbarna/gcn.nasa.gov/blob/6d1838715fdd7b4da9b28ef040fb4e1ddfc856de/app/routes/circulars/circulars.server.ts#L70)), though I see why it's better to only keep it as one type. Do you think it would be better to keep it as the string up to the point of defining it in the search function?

File: app/routes/circulars/circulars.server.ts (hunk 4 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.server.ts:102]: Looking through some docs ([1](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-source-field.html)) ([2](https://opster.com/guides/elasticsearch/glossary/elasticsearch-source/)), I think I implemented the syntax slightly incorrectly, but setting it to false seems to break the circulars indexing on initial load. I probably need to look this over a bit more to understand what's happening, but altering it to the following seems to work (doesn't fix the query submission issue that I'm encountering though):

```
      _source: {
        includes: ['subject', 'createdOn'],
      }
```

File: app/routes/circulars/circulars.server.ts (hunk 5 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.server.ts]: yeah, not sure how this happened, I think my fork may have been out of date when I created this branch or something along those lines. Resolved in [adf6556](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1026/commits/adf6556f2ea022c03eea7237e9efd56f2d737254)

File: app/routes/circulars/circulars.server.ts (hunk 3 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.server.ts]: Made this change here and in relevant functions in [67335f3](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1026/commits/67335f39365fe353b6c9a2d5b513efe572d1a73e). I think I fixed search being passed the default values (Infinity and -Infinity), but need to check if providing Infinity as a string to formatDateEpoch will return Infinity as a number

File: app/routes/circulars/circulars.lib.ts (hunk 1 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.lib.ts]: I added a function for YYYY-MM-DD in 0e2c267. Given the goal to have the input and url string be human readable in  iso8601, this function might end up being unneeded, as the start and end dates should just be passed as strings at all points and only converted to unix time inside the search function. That being said, looking around a bit more, the default infinity values don't seem to interact well when using the Date objects. One solution to that could be to change it so the default start and end dates are undefined and at the point of conversion inside the search, they are set to infinity and negative infinity if they're undefined (similar to how query is defined when there is no query), though this would require tweaking the declaration of their types in the various functions.

Comment (lpsinger) [reply app/routes/circulars/circulars.lib.ts]: Right. We don't need the function to convert from epoch to ISO. We only need the function to convert from ISO to epoch.

I agree: use undefined, not infinity.

File: app/routes/circulars/circulars.lib.ts [hunk 6]
```diff
@@ -76,6 +76,28 @@ export function formatDateISO(date: number) {
...
+    day: '2-digit',
+  })
+  return dateStr.replace(/\//g, '-') // replace / with -
+}
+/** convert a ms value to YYYY-MM-DD format */
+export function formatDateYYYYMMDD(date: number) {
+  const dateObj = new Date(date)
+  var dateStr = dateObj.toISOString().split('T')[0]
+  return dateStr
+}
```
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:94]: Remove, unused:

```suggestion
```

File: app/routes/circulars/circulars.lib.ts [hunk 7]
```diff
@@ -76,6 +76,28 @@ export function formatDateISO(date: number) {
...
+
+/** convert a date in format mm-dd-yyyy (or YYYY-MM_DD) to ms since 01/01/1970 */
+export function formatDateEpoch(date: string) {
+  return new Date(date).getTime()
+}
```
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:99]: Take care: when you search for records with dates between `yyyy-mm-dd` (lower limit) and `YYYY-MM-DD` (upper limit) that you are searching between the times `yyyy-mm-ddT00:00:00Z` and `YYYY-MM-DDT23:59:59Z` (so that hits include the whole day that comprises the upper end of the interval).

File: app/routes/circulars/index.tsx [hunk 8]
```diff
@@ -151,12 +184,20 @@ export default function () {
...
+
   let searchParamsString = searchParams.toString()
   if (searchParamsString) searchParamsString = `?${searchParamsString}`
+  const combinedQuery = { query, page: page, startDate, endDate }
+  console.log('default function combinedQuery', combinedQuery)
 
-  const [input, setInput] = useState(query)
-  const clean = input === query
+  const [input, setInput] = useState(combinedQuery)
+  const clean = input === combinedQuery
```
Comment (lpsinger) [inline app/routes/circulars/index.tsx:198]: My advice: don't create states that are composite objects. Have a separate state variable for each input field, like this:

```ts
const [inputQuery, setInputQuery] = useState(query)
const [inputDateGte, setInputDateGte] = useState(gte)
const [inputDateLte, setInputDateLte] = useState(lte)
const clean = inputQuery === query && inputDateGte === gte && inputDateLte === lte
...
```

The reason is that there are [many subtleties around properly updating React states that are objects](https://react.dev/learn/updating-objects-in-state).

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by lpsinger** [comment]
@dakota002, while I am out next week, would you please mentor @tylerbarna and help him to work out the remaining kinks in this PR?

File: app/routes/circulars/index.tsx (hunk 8 above)
Comment (tylerbarna) [reply app/routes/circulars/index.tsx:198]: this actually fixed the issue with the loading not working I think

File: app/routes/circulars/circulars.lib.ts (hunk 6 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.lib.ts]: removed in fbce467

File: app/routes/circulars/circulars.lib.ts (hunk 1 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.lib.ts]: > Right. We don't need the function to convert from epoch to ISO. We only need the function to convert from ISO to epoch.
> 
> I agree: use undefined, not infinity.

made the change so they're undefined in 64a34b2.

File: app/routes/circulars/circulars.server.ts (hunk 5 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.server.ts]: with most recent commit (64a34b2), double checked that the formatDateEpoch function works inside of the search function to limit the range of circulars by date. Manually changed the definition of startTime and endTime constants so instead of referencing startDate and endDate, they were hard-coded strings in the YYYY-MM-DD format. With this, the search function successfully limited the circulars to the provided date range (prior to the commit, I reverted the hard coded date range back to the constants). The backend for the search function seems to be functional as of now.

**Comment by tylerbarna** [comment]
with most recent commit (https://github.com/nasa-gcn/gcn.nasa.gov/commit/64a34b23330bd96b2584383d677af3df3f8e81d2), double checked that the formatDateEpoch function works inside of the search function to limit the range of circulars by date. Manually changed the definition of startTime and endTime constants so instead of referencing startDate and endDate, they were hard-coded strings in the YYYY-MM-DD format. With this, the search function successfully limited the circulars to the provided date range (prior to the commit, I reverted the hard coded date range back to the constants). The backend for the search function seems to be functional as of now.

**Comment by tylerbarna** [comment]
note that, while the backend does not have a frontend, the [definition of the start and end date input constants](https://github.com/tylerbarna/gcn.nasa.gov/blob/8999d1529269a10eb3d313d82168d322f188690b/app/routes/circulars/index.tsx#L197) are commented out so commits can pass the husky pre-commit hook (when uncommented, they are defined but not referenced). With the frontend, they will be referenced inside of the date range picker elements similar to the way [inputQuery](https://github.com/tylerbarna/gcn.nasa.gov/blob/8999d1529269a10eb3d313d82168d322f188690b/app/routes/circulars/index.tsx#L196) is referenced [in the search box element](https://github.com/tylerbarna/gcn.nasa.gov/blob/8999d1529269a10eb3d313d82168d322f188690b/app/routes/circulars/index.tsx#L230)

**Comment by tylerbarna** [comment]
I noticed while looking over the work for this PR, but the changes made in #967 are such that there won't need to be any additional changes for startDate, endDate, and any future URLSearchParameters to be preserved when using the back button after navigating to a specific circular.

File: app/routes/circulars/circulars.lib.ts (hunk 7 above)
Comment (tylerbarna) [reply app/routes/circulars/circulars.lib.ts:82]: I made a small change in 55e52d9 so the endDate, when defined, should include circulars recorded on the date of the endDate. I did this by just adding the number of milliseconds in a day to the endTime constant, which is in unix format,  when it's defined inside of the search function. Wasn't sure if that's sufficient or if it would it be better to look for a more elegant implementation.

**Comment by tylerbarna** [comment]
@dakota002 After applying some of Leo's requested changes, the date range backend seems to be working without causing any issues with the standard query search. Still need to remove any console.log calls in index.tsx that I was using for debugging.

**Comment by tylerbarna** [comment]
> > @dakota002 After applying some of Leo's requested changes, the date range backend seems to be working without causing any issues with the standard query search. Still need to remove any console.log calls in index.tsx that I was using for debugging.
> 
> Awesome, let me know when you want me to take a look at the PR

yeah, I think it's probably ready for review

File: app/routes/circulars/index.tsx [hunk 9]
```diff
@@ -151,12 +183,18 @@ export default function () {
...
+
   let searchParamsString = searchParams.toString()
   if (searchParamsString) searchParamsString = `?${searchParamsString}`
 
-  const [input, setInput] = useState(query)
-  const clean = input === query
+  const [inputQuery, setInputQuery] = useState(query)
+  // const [inputDateGte, setInputDateGte] = useState(startDate)
+  // const [inputDateLte, setInputDateLte] = useState(endDate)
+  const clean = inputQuery === query //&& inputDateGte === startDate && inputDateLte === endDate
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:197]: These commented out lines can be deleted

File: app/routes/circulars/index.tsx [hunk 10]
```diff
@@ -34,13 +34,25 @@ import searchImg from 'app/theme/img/usa-icons-bg/search--white.svg'
 const limit = 100
 
 export async function loader({ request: { url } }: DataFunctionArgs) {
+  console.log('loader url', url)
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:37]:
```suggestion
```

File: app/routes/circulars/index.tsx [hunk 11]
```diff
@@ -34,13 +34,25 @@ import searchImg from 'app/theme/img/usa-icons-bg/search--white.svg'
...
   const { searchParams } = new URL(url)
+  console.log('loader search params', searchParams)
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:39]:
```suggestion
```

File: app/routes/circulars/index.tsx [hunk 12]
```diff
@@ -34,13 +34,25 @@ import searchImg from 'app/theme/img/usa-icons-bg/search--white.svg'
...
   const query = searchParams.get('query') || undefined
   if (query) {
     await circularRedirect(query)
   }
+  const startDate = searchParams.get('startDate') || undefined
+  const endDate = searchParams.get('endDate') || undefined
   const page = parseInt(searchParams.get('page') || '1')
-  const results = await search({ query, page: page - 1, limit })
+  console.log('loader page', page)
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:47]:
```suggestion
```

File: app/routes/circulars/index.tsx [hunk 13]
```diff
@@ -53,22 +65,42 @@ export async function action({ request }: DataFunctionArgs) {
...
+  query?: string,
+  startDate?: string,
+  endDate?: string
+) {
+  const searchParams = new URLSearchParams({ page: page.toString() || '1' })
   if (query) searchParams.set('query', query)
+  searchParams.set('page', page.toString() || '1')
+  if (startDate) searchParams.set('startDate', startDate)
+  if (endDate) searchParams.set('endDate', endDate)
+  console.log('getPageLink search params', searchParams.toString())
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:79]:
```suggestion
```

File: app/routes/circulars/index.tsx [hunk 14]
```diff
@@ -53,22 +65,42 @@ export async function action({ request }: DataFunctionArgs) {
...
+  endDate?: string
 }) {
   const pages = usePagination({ currentPage: page, totalPages })
+  console.log(
+    'Pagination search params',
+    query?.toString(),
+    startDate?.toString(),
+    endDate?.toString()
+  )
+  console.log('Pagination page', page)
```
Comment (dakota002) [inline app/routes/circulars/index.tsx:103]:
```suggestion
```

**Review by dakota002 — CHANGES_REQUESTED** [review]
Looks good, but these are just some cleanup requests.  Please remove commented code and console logs

File: app/routes/circulars/index.tsx (hunk 9 above)
Comment (tylerbarna) [reply app/routes/circulars/index.tsx]: these are for the frontend, but yeah, probably best to leave them out until the actual frontend PR

**Comment by tylerbarna** [comment]
> Looks good, but these are just some cleanup requests. Please remove commented code and console logs

I've implemented all the requested changes

**Review by dakota002 — APPROVED** [review]
Looks good to me!

**Comment by tylerbarna** [comment]
@dakota002 I believe I squashed the commits down

**Comment by tylerbarna** [comment]
> @dakota002 I believe I squashed the commits down

in the future, is it better to squash the commits in the same PR/branch, or should I transfer everything to a new branch/PR with a single commit?

**Comment by dakota002** [comment]
> > @dakota002 I believe I squashed the commits down
> 
> in the future, is it better to squash the commits in the same PR/branch, or should I transfer everything to a new branch/PR with a single commit?

You can do it all on the same PR/branch. I've had a few cases where I've needed to do it multiple times over on the same branch to avoid large conflicts
